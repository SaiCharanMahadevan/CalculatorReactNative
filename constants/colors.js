const COLORS = {
  BLACK: '#1C1C1C',
  WHITE: '#FFFFFF',
  LIGHT_PINK: '#FEDBDC',
  PRIMARY_PINK: '#EE2F5E',
  LIGHT_GREY: '#E8EAEF',
  PRIMARY_GREY: '#BCCOCE',
  DARK_GREY: '#474449',
}

export default COLORS;