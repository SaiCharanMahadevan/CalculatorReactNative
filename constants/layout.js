const LAYOUT = {
  GUTTER: 16,
  GAP: 12,
  TEXT_PADDING: 4,
  BUTTON_HEIGHT: 75,
  BUTTON_WIDTH: 75,
};

export default LAYOUT;