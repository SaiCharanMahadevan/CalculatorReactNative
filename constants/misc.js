export const BUTTONS_OPERATORS = [
  'รท',
  'x',
  '+',
  '-'
];
export const BUTTONS_NUMBERS = [
  ['7', '8', '9'],
  ['4', '5', '6'],
  ['1', '2', '3']
];
export const CLEAR = 'DELETE';
export const ZERO = '0';
export const EQUALS = '=';
export const ERROR_MESSAGE = `Can't do this!`;