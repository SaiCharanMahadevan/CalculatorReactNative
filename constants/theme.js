import COLORS from './colors';
import LAYOUT from './layout';

const THEME = {
  COLORS,
  LAYOUT,
  FONT: {
    SIZES: {
      small: 18,
      normal: 25,
      medium: 40,
      large: 60,
    },
    WEIGHTS: {
      normal: 400,
      semibold: 500,
      bold: 700,
    }
  },
};

export default THEME;