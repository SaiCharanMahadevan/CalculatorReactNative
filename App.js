import React, { Fragment, PureComponent } from 'react';
import { SafeAreaView, Platform } from 'react-native';
import styled, { ThemeProvider } from 'styled-components/native';

import {
  Button, Col, FlatButton, FlexRow, FlexView, Text, GapSeparator,
} from 'components';
import { InputButtonsContainer, OutputContainer } from 'containers'
import THEME from 'constants/theme';
import LAYOUT from 'constants/layout';;
import { CLEAR, ERROR_MESSAGE, EQUALS } from 'constants/misc';

const Wrapper = styled.SafeAreaView`
  flex: 1;
  backgroundColor: ${({ theme: { COLORS } }) => COLORS.DARK_GREY};
  paddingTop: ${({ isAndroid }) => (isAndroid ? 25 : 0)};
`;
const LargeGap = styled(FlexView)`
  flex: 1;
`
const MAX_LENGTH = 57
export default class App extends PureComponent {
  constructor(props){
    super(props);
    this.state = {
      result: '',
      expression: '',
    }
  }

  _escapeRegExp = (str) => {
    return str.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
  }

     // Calculate the value of the expression after checking its validity
  _evaluate = () => {
    const { expression } = this.state;
    let temp = expression;
    try{
      // Evalutate if operators exist in the expression. Edge Case - `0x` return false (HexDecimal equivalent)
      if(isNaN(temp) || (temp[0] === '0' && temp[1] === 'x')){
        let strTemp = temp.replace(new RegExp(this._escapeRegExp('รท'), 'g'), '/');
        strTemp = strTemp.replace(new RegExp(this._escapeRegExp('x'), 'g'), '*');
        temp = eval(strTemp);
        if(temp%1 > 0) {
          temp = temp.toFixed(5)
        }

        if(temp === Infinity) {
          temp = ERROR_MESSAGE
        }
      }

      this.setState({
        result: temp.toString(),
      })
    } catch(exception){
        console.log('exception: ' + exception);
    }
  }

  concatToOutput = (value) => {
    const { expression } = this.state;
    if(expression.length >= MAX_LENGTH || ((expression === '') && isNaN(value))){
      return;
    }
    
    if(expression === ''){
      this.setState({expression: value})
    } else{
      this.setState({expression: `${expression}${value}`})
    }

    this._evaluate();
  }

  // Handle button press actions
  handleEvent = (value) => {
    const { expression, result } = this.state;
    switch(value) {
      case CLEAR:
        if(result === ERROR_MESSAGE) {
          this.setState({
            expression: '',
          });
        } else {
          this.replaceLastIndex('');
        }
        break;
      case EQUALS:
        this.updateAnswer();
        break;
      default:
        this.updateExpression(value);
        break;
    }
  }

  // Handle long press actions
  handleLongPress = (value) => {
    if(value === CLEAR) {
      this.reset();
    }
  }

  // Removes the last index of the expression.
  replaceLastIndex = (value) => {
    const { expression } = this.state;
    let newExpression = expression.replace(/.$/, value)
    
    this.setState({
      expression: expression.replace(/.$/, value)
    })
  
    this._evaluate();
  }

  // Resets to the start state.
  reset = () => {
    this.setState({
      result: '',
      expression: ''
    })
  }

  updateAnswer = () => {
    const { result } = this.state;
    this.setState({
      expression: result,
      result: '',
    })
  }

  // Forms the latest expression ready for evaluation
  updateExpression = (value) => {
    const { expression } = this.state;
    if(expression === ERROR_MESSAGE) {
      this.setState({
        expression: '',
      });
    }
    let strLastChar = expression.slice(expression.length-1);
    if(isNaN(strLastChar) && isNaN(value)){
      this.replaceLastIndex(value);
    } else {
      this.concatToOutput(value);
    }
  }

  render() {
    const { expression, result } = this.state;
    return (
      <ThemeProvider theme={THEME}>
        <Wrapper isAndroid={Platform.OS === 'android'}>
          <FlexView flex={1}>
            <LargeGap />
            <OutputContainer expression={expression} result={result}/>
            <InputButtonsContainer onButtonPress={this.handleEvent} onLongPress={this.handleLongPress} />
          </FlexView>
        </Wrapper>
      </ThemeProvider>
    );
  }
  
}
