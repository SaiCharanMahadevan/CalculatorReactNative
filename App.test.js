import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components/native';
import { CLEAR, ERROR_MESSAGE, EQUALS } from 'constants/misc';

import THEME from 'constants/theme';
import App from './App';

describe('App UI: ', () => {
  test('renders the Calculator', () => {
    const wrapper = shallow(
      <App />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('App Functionality: ', () => {
  test('updates the expression and result by concatinating the expression when a number is pressed ', () => {
    const wrapper = shallow(
      <App />,
    );
  
    wrapper.instance().handleEvent('1')
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('removes the last input when CLEAR is pressed', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '12' })
    wrapper.instance().handleEvent(CLEAR)
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('resets input when CLEAR is long pressed', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '2123+2323' })
    wrapper.instance().reset()
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('updates the expression with the result when EQUALS is pressed', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '2123+2323' })
    wrapper.instance()._evaluate()
    wrapper.instance().updateAnswer()
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('does not update the expression if the last and next inputs are invalid ', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '1+' })
    wrapper.instance().handleEvent('+')
    expect(wrapper.instance().state).toMatchSnapshot();
    wrapper.setState({ expression: '' })
    wrapper.instance().handleEvent('+')
    expect(wrapper.instance().state).toMatchSnapshot();
  });


  test('updates the expression by replacing the operator if the last input was an operator ', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '1+' })
    wrapper.instance().updateExpression('*')
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('clears the expression after any next input when the result displays the error message', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: ERROR_MESSAGE })
    wrapper.instance().handleEvent('*')
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  test('updates the expression to handle upto 5 decimals', () => {
    const wrapper = shallow(
      <App />,
    );
    
    wrapper.setState({ expression: '9รท' })
    wrapper.instance().handleEvent('7')
    expect(wrapper.instance().state).toMatchSnapshot();
  });
});
