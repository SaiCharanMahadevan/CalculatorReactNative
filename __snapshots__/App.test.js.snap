// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Functionality:  clears the expression after any next input when the result displays the error message 1`] = `
Object {
  "expression": "",
  "result": "",
}
`;

exports[`App Functionality:  does not update the expression if the last and next inputs are invalid  1`] = `
Object {
  "expression": "1+",
  "result": "",
}
`;

exports[`App Functionality:  does not update the expression if the last and next inputs are invalid  2`] = `
Object {
  "expression": "",
  "result": "",
}
`;

exports[`App Functionality:  removes the last input when CLEAR is pressed 1`] = `
Object {
  "expression": "1",
  "result": "1",
}
`;

exports[`App Functionality:  resets input when CLEAR is long pressed 1`] = `
Object {
  "expression": "",
  "result": "",
}
`;

exports[`App Functionality:  updates the expression and result by concatinating the expression when a number is pressed  1`] = `
Object {
  "expression": "1",
  "result": "1",
}
`;

exports[`App Functionality:  updates the expression by replacing the operator if the last input was an operator  1`] = `
Object {
  "expression": "1*",
  "result": "",
}
`;

exports[`App Functionality:  updates the expression to handle upto 5 decimals 1`] = `
Object {
  "expression": "9รท7",
  "result": "1.28571",
}
`;

exports[`App Functionality:  updates the expression with the result when EQUALS is pressed 1`] = `
Object {
  "expression": "4446",
  "result": "",
}
`;

exports[`App UI:  renders the Calculator 1`] = `
<ThemeProvider
  theme={
    Object {
      "COLORS": Object {
        "BLACK": "#1C1C1C",
        "DARK_GREY": "#474449",
        "LIGHT_GREY": "#E8EAEF",
        "LIGHT_PINK": "#FEDBDC",
        "PRIMARY_GREY": "#BCCOCE",
        "PRIMARY_PINK": "#EE2F5E",
        "WHITE": "#FFFFFF",
      },
      "FONT": Object {
        "SIZES": Object {
          "large": 60,
          "medium": 40,
          "normal": 25,
        },
        "WEIGHTS": Object {
          "bold": 700,
          "normal": 400,
          "semibold": 500,
        },
      },
      "LAYOUT": Object {
        "BUTTON_HEIGHT": 75,
        "BUTTON_WIDTH": 75,
        "GAP": 12,
        "GUTTER": 16,
        "TEXT_PADDING": 4,
      },
    }
  }
>
  <Styled(SafeAreaView)
    isAndroid={false}
  >
    <Styled(View)
      flex={1}
    >
      <Styled(Styled(View)) />
      <OutputContainer
        expression=""
        result=""
      />
      <InputButtonsContainer
        onButtonPress={[Function]}
        onLongPress={[Function]}
      />
    </Styled(View)>
  </Styled(SafeAreaView)>
</ThemeProvider>
`;
